tuned/plugins/plugin_irq.py

```diff
@@ -0,0 +1,167 @@
+from . import hotplug
+from .decorators import *
+import tuned.consts as consts
+import tuned.logs
+from tuned.utils.commands import commands
+
+import errno
+import os
+
+log = tuned.logs.get()
+
+class IrqPlugin(hotplug.Plugin):
+
+	def __init__(self, monitor_repository, storage_factory, hardware_inventory, device_matcher, device_matcher_udev, plugin_instance_factory, global_cfg, variables):
+		super(IrqPlugin, self).__init__(monitor_repository, storage_factory, hardware_inventory, device_matcher, device_matcher_udev, plugin_instance_factory, global_cfg, variables)
+		self._irq_original = {}
+		self._irq_unchangeable = []
+
+	def _init_devices(self):
+		self._devices_supported = True
+		self._free_devices = set()
+		self._assigned_devices = set()
+		for i in os.listdir("/proc/irq"):
+			p = os.path.join("/proc/irq", i)
+			if os.path.isdir(p) and i.isdigit():
+				self._free_devices.add("irq%s" % i)
+
+	@classmethod
+	def _get_config_options(self):
+		return {
+			"affinity": "",
+			"default_affinity": "calc",
+		}
+
+	def _instance_init(self, instance):
+		instance._has_static_tuning = True
+		instance._has_dynamic_tuning = False
+		instance._is_main_instance = instance.name == instance.plugin.name
```

Comment:
I wouldn't do this. Instance name can be arbitrary string.  So maybe you want the first instance?

Do we really need `default_affinity`?

E.g.:
```
[irq]
affinity=1,2,3
```

Or:
```
[irq_a]
type=irq
devices=1,2
affinity=1,2

[irq_b]
devices=*
affinity=3
```

If I am not missing anything, I think we could live without it and simplify the code.

Reply:
This `default_affinity` is not a default used by TuneD/plugin_irq, but it's actually set in the kernel (`/proc/irq/default_smp_affinity`) and affects newly appearing irqs, even before TuneD sees them. So yes, I'd say we need the setting, and it only makes sense to set it once.

But yes, instead of using the name it's probably better to take the "first instance" approach used by the `cpu` plugin.